File: src/ros2_bridge/mappers/sensor_msgs/joy.rs
//! Typed mapper for `sensor_msgs/msg/Joy`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn joy_to_bus(msg: ros_env::sensor_msgs::msg::Joy) -> crate::sensor_msgs::msg::v1::Joy {
    crate::sensor_msgs::msg::v1::Joy {
        header: Some(crate::ros2_bridge::mappers::std_msgs::header::header_to_bus(msg.header)),
        axes: crate::ros2_bridge::mappers::convert::f32_seq(msg.axes),
        buttons: crate::ros2_bridge::mappers::convert::i32_seq(msg.buttons),
    }
}

pub(crate) fn joy_to_ros(bus: crate::sensor_msgs::msg::v1::Joy) -> ros_env::sensor_msgs::msg::Joy {
    ros_env::sensor_msgs::msg::Joy {
        header: crate::ros2_bridge::mappers::std_msgs::header::header_to_ros(bus.header.unwrap_or_default()),
        axes: bus.axes,
        buttons: bus.buttons,
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct SensorMsgsJoyMapper;

impl TypedTopicMapper for SensorMsgsJoyMapper {
    type Ros = ros_env::sensor_msgs::msg::Joy;
    type Bus = crate::sensor_msgs::msg::v1::Joy;

    fn type_name(&self) -> &'static str {
        "sensor_msgs/msg/Joy"
    }

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(joy_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(joy_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/sensor_msgs/joy.py
"""Generated mapper for `sensor_msgs/msg/Joy`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.std_msgs.header import header_to_bus, header_to_ros

_BusMsg = None


def _bus_cls():
    global _BusMsg
    if _BusMsg is None:
        from robot_bus.sensor_msgs.msg.v1 import Joy as BusMsg

        _BusMsg = BusMsg
    return _BusMsg


def joy_to_bus(msg):
    BusMsg = _bus_cls()
    bus = BusMsg()
    bus.header.CopyFrom(header_to_bus(msg.header))
    bus.axes.extend(list(msg.axes))
    bus.buttons.extend(list(msg.buttons))
    return bus


def joy_to_ros(bus):
    from sensor_msgs.msg import Joy as RosMsg

    out = RosMsg()
    out.header = header_to_ros(bus.header)
    out.axes = list(bus.axes)
    out.buttons = list(bus.buttons)
    return out


class SensorMsgsJoyMapper:
    _ros_type = None

    def ros_msg_type(self):
        cls = type(self)
        if cls._ros_type is None:
            from sensor_msgs.msg import Joy as RosMsg

            cls._ros_type = RosMsg
        return cls._ros_type

    def ros_to_bus(self, msg) -> bytes:
        return joy_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        BusMsg = _bus_cls()
        bus = BusMsg()
        bus.ParseFromString(payload)
        return joy_to_ros(bus)
